File: backend/app/routers/analyze.py
"""Face analysis endpoint."""

from __future__ import annotations

import time
from typing import Optional, Union

from fastapi import APIRouter, File, Request, UploadFile
from fastapi.responses import JSONResponse

from app.models.schemas import (
    AnalyzeData,
    AnalyzeResponse,
    ErrorCodes,
    ErrorDetail,
    ErrorResponse,
    ResponseMeta,
)
from app.services.face_detection import get_face_detection_service
from app.utils.image import ImageValidationError, validate_image

router = APIRouter(tags=["analyze"])


@router.post("/analyze", response_model=None)
async def analyze_face(
    request: Request,
    image: Optional[UploadFile] = File(None),
):
    """
    Analyze a face in an uploaded image.

    Detects face and extracts landmarks using MediaPipe Face Mesh.

    Args:
        request: FastAPI request object
        image: Uploaded image file (JPEG or PNG, max 10MB)

    Returns:
        Face analysis results including landmarks and face region
    """
    start_time = time.time()
    request_id = getattr(request.state, "request_id", "unknown")

    # Validate image file is provided
    if image is None or image.filename == "":
        return JSONResponse(
            status_code=400,
            content=ErrorResponse(
                error=ErrorDetail(
                    code=ErrorCodes.VALIDATION_ERROR,
                    message="Image file is required",
                )
            ).model_dump(),
        )

    try:
        # Read image data
        image_data = await image.read()

        if len(image_data) == 0:
            return JSONResponse(
                status_code=400,
                content=ErrorResponse(
                    error=ErrorDetail(
                        code=ErrorCodes.VALIDATION_ERROR,
                        message="Image file is required",
                    )
                ).model_dump(),
            )

        # Validate and load image
        image_format, cv2_image = validate_image(image_data)

        # Detect faces
        face_service = get_face_detection_service()
        result = face_service.detect(
            cv2_image,
            image_format=image_format,
            require_single_face=True,
        )

        # Calculate processing time
        processing_time_ms = (time.time() - start_time) * 1000

        return AnalyzeResponse(
            data=AnalyzeData(
                face_detected=result.face_detected,
                face_count=result.face_count,
                face_region=result.face_region,
                landmarks=result.landmarks,
                image_info=result.image_info,
            ),
            meta=ResponseMeta(
                request_id=request_id,
                processing_time_ms=processing_time_ms,
            ),
        )

    except ImageValidationError as e:
        # Map error code to HTTP status
        status_code_map = {
            ErrorCodes.IMAGE_TOO_LARGE: 413,
            ErrorCodes.INVALID_IMAGE_FORMAT: 400,
            ErrorCodes.FACE_NOT_DETECTED: 400,
            ErrorCodes.MULTIPLE_FACES: 400,
        }
        status_code = status_code_map.get(e.code, 400)

        return JSONResponse(
            status_code=status_code,
            content=ErrorResponse(
                error=ErrorDetail(
                    code=e.code,
                    message=e.message,
                    details=e.details,
                )
            ).model_dump(),
        )

    except Exception as e:
        # Log error (in production, use proper logging)
        print(f"Analyze error: {e}")

        return JSONResponse(
            status_code=500,
            content=ErrorResponse(
                error=ErrorDetail(
                    code=ErrorCodes.PROCESSING_ERROR,
                    message="An unexpected error occurred during image analysis",
                )
            ).model_dump(),
        )

File: frontend/src/lib/api/analyze.ts
/**
 * 顔分析API
 */

import { apiPostFormData } from './client';
import { AnalyzeData } from './types';

/**
 * 画像から顔を分析する
 *
 * @param image - 分析対象の画像ファイル（JPEG/PNG、最大10MB）
 * @returns 顔分析結果（顔領域、ランドマーク、画像情報）
 * @throws ApiError - 顔未検出、複数顔検出、画像フォーマットエラーなど
 *
 * @example
 * ```typescript
 * const fileInput = document.querySelector<HTMLInputElement>('#file-input');
 * const file = fileInput?.files?.[0];
 * if (file) {
 *   try {
 *     const result = await analyzeImage(file);
 *     if (result.face_detected) {
 *       console.log('Face found at:', result.face_region);
 *       console.log('Landmarks:', result.landmarks?.length);
 *     }
 *   } catch (error) {
 *     if (error instanceof ApiError && error.code === 'FACE_NOT_DETECTED') {
 *       alert('顔を検出できませんでした');
 *     }
 *   }
 * }
 * ```
 */
export async function analyzeImage(image: File): Promise<AnalyzeData> {
  const formData = new FormData();
  formData.append('image', image);

  return apiPostFormData<AnalyzeData>('/api/v1/analyze', formData);
}
